include/up-cpp/datamodel/validator/Uuid.h

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-FileType: SOURCE
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_DATAMODEL_VALIDATOR_UUID_H
+#define UP_CPP_DATAMODEL_VALIDATOR_UUID_H
+
+#include <uprotocol/v1/uuid.pb.h>
+
+#include <chrono>
+#include <cstdint>
+#include <stdexcept>
+
+// @brief Collection of functions for validating and inspecting UUIDs.
+namespace uprotocol::datamodel::validator::uuid {
+
+/// @brief This exception indicates that a UUID object was provided that
+///        did not contain valid UUID data.
+struct InvalidUuid : public std::invalid_argument {
+	// Inherit constructors
+	using std::invalid_argument::invalid_argument;
+
+	InvalidUuid(InvalidUuid&&) noexcept;
+	InvalidUuid& operator=(InvalidUuid&&) noexcept;
+
+	InvalidUuid(const InvalidUuid&);
+	InvalidUuid& operator=(const InvalidUuid&);
+};
+
+/// @name Validity checks
+/// @{
+/// @brief Checks if the provided UUID contains valid v6 or v8 UUID data.
+/// @returns True if the UUID has valid UUID data, false otherwise.
+bool isUuid(v1::UUID);
+
+/// @brief Checks if the provided UUID contains valid v6 UUID data.
+/// @returns True if the UUID has valid UUID v6 data, false otherwise.
+bool isV6(v1::UUID);
+
+/// @brief Checks if the provided UUID contains valid v8 (uProtocol)
+///        UUID data.
+/// @returns True if the UUID has a valid uProtocol UUID, false otherwise.
+bool isUProtocol(v1::UUID);
+
+/// @breif Checks if the provided UUID has expired based on the given TTL.
+/// @throws InvalidUuid if the UUID does not contain valid UUID data
+/// @returns True if the difference between the current system time and
+///          the the timestamp in the UUID is greater than the TTL.
+bool isExpired(v1::UUID uuid, std::chrono::milliseconds ttl);
+/// @}
+
+/// @name Inspection utilities
+/// @{
+/// @breif Gets the version field from a UUID object
+/// @throws InvalidUuid if the UUID does not contain valid UUID data
```

Comment:
Interesting. So there's not a way in C++ to tell which exceptions a function can throw other than by:

* reading the code or
* relying on documentation

?

Reply:
You are correct

Reply:
Huh, I sorta knew this was the case, but hadn't used exceptions in a production environment before.

My previous project in C++ used a Rust-style Result type.

I suppose may just be my lack of familiarity / comfort with exceptions.

But if this is considered normal, the I'll accept that.